server/src/test/java/org/opensearch/action/pagination/ShardPaginationStrategyTests.java

```diff
@@ -0,0 +1,400 @@
+            // deleting test-index-7 & 10 more random indices after 11th call. By that time, shards for first 6
+            // indices would have been completely fetched. 11th call, would fetch first 3 shardsIDs for test-index-7 and
+            // if it gets deleted, rest 2 shardIDs for it should not appear.
+            if (numPages == 11) {
+                clusterState = deleteIndexFromClusterState(clusterState, fixedIndexNumToDelete);
+
+                deletedIndices = indexNumberList.subList(20, indexNumberList.size());
+                Collections.shuffle(deletedIndices, getRandom());
+                for (int pos = 0; pos < numIndicesToDelete; pos++) {
+                    clusterState = deleteIndexFromClusterState(clusterState, deletedIndices.get(pos));
+                }
+            }
+            // creating 5 indices after 5th call
+            if (numPages == 5) {
+                for (int indexNumber = totalIndices + 1; indexNumber <= totalIndices + numIndicesToCreate; indexNumber++) {
+                    clusterState = addIndexToClusterState(clusterState, indexNumber, numShardsForNewIndices, numReplicasForNewIndices);
+                }
+            }
+            assertTrue(paginationStrategy.getRequestedEntities().size() <= pageSize);
+            shardRoutings.addAll(paginationStrategy.getRequestedEntities());
+            indicesFetched.addAll(paginationStrategy.getRequestedIndices());
+        } while (Objects.nonNull(requestedToken));
+
+        assertEquals(totalIndices + numIndicesToCreate - numIndicesToDelete, indicesFetched.size());
+
+        // finalTotalShards = InitialTotal - 2ShardIdsForIndex7 - ShardsFor10RandomlyDeletedIndices + ShardsForNewlyCreatedIndices
+        final int totalShards = totalInitialShards - 2 * (DEFAULT_NUMBER_OF_REPLICAS + 1) - numIndicesToDelete * DEFAULT_NUMBER_OF_SHARDS
+            * (DEFAULT_NUMBER_OF_REPLICAS + 1) + numIndicesToCreate * numShardsForNewIndices * (numReplicasForNewIndices + 1);
+        assertEquals(totalShards, shardRoutings.size());
+
+        // deleted test-index-7, should appear in the response shards and indices
+        assertTrue(indicesFetched.contains(TEST_INDEX_PREFIX + 7));
+        assertEquals(
+            shardRoutings.stream().filter(shard -> shard.getIndexName().equals(TEST_INDEX_PREFIX + 7)).count(),
+            3 * (DEFAULT_NUMBER_OF_REPLICAS + 1)
+        );
+        // none of the randomly deleted index should appear in the list of fetched indices
+        for (int deletedIndexPos = 0; deletedIndexPos < numIndicesToDelete; deletedIndexPos++) {
+            String indexName = TEST_INDEX_PREFIX + deletedIndices.get(deletedIndexPos);
+            assertFalse(indicesFetched.contains(indexName));
+            assertEquals(shardRoutings.stream().filter(shard -> shard.getIndexName().equals(indexName)).count(), 0);
+
+        }
+
+        // all the newly created indices should be present in the list of fetched indices
```

Comment:
- Do we need to also assert the order in which they are returned (or) the page they were part of when the response was returned ?